OpenBSD installer / instructions

Any chance of getting a binary installer for OpenBSD (current, AMD64) or instructions on
how to get it working from source? I develop Haskell on OpenBSD and would like to use
stack.

@mrijkeboer I've tried the `ghc-7.10.2-openbsd-x86_64-5.8.tar.bz2` bindist using the `configure` defaults, and run into some trouble:

* Compiling a simple program doesn't work with default options.  E.g.:

        $ cat >>hello.hs
        main = putStrLn "Hello, world"
        $ ghc hello.hs
        [1 of 1] Compiling Main             ( hello.hs, hello.o )
        Linking hello ...
        /usr/bin/ld: /home/manny/.local/ghc-7.10.2.20151026/lib/ghc/rts/libCffi.a(prep_cif.o): relocation R_X86_64_32 against `a local symbol' can not be used when making a shared object; recompile with -fPIC
        /home/manny/.local/ghc-7.10.2.20151026/lib/ghc/rts/libCffi.a: could not read symbols: Bad value
        collect2: ld returned 1 exit status

* `runghc` doesn't seem to work.  E.g.:

        $ runghc hello.hs
        <command line>: not built for interactive use

* There's no `haddock` binary included, which means cabal-install's `bootstrap.sh` doesn't run
* The version reported by `ghc --version` is `7.10.2.20151026` rather than `7.10.2`

I downloaded the GHC source (7.10.2) and can run ./configure without problems. When building, ghc-stage1 is build successful, but when building the base library I get the following error:
```
checking value of O_BINARY... 0
checking for library containing iconv... no
configure: error: iconv is required on non-Windows platforms
libraries/base/ghc.mk:4: recipe for target 'libraries/base/dist-install/package-data.mk' failed
gmake[1]: *** [libraries/base/dist-install/package-data.mk] Error 1
Makefile:71: recipe for target 'build' failed
gmake: *** [build] Error 2
```

The strange thing is that libiconv is installed in /usr/local/.

@Blaisorblade It seems that the GHC 8.0.2 is build for OpenBSD-current or at least it works:

```
$ sysctl kern.version
kern.version=OpenBSD 6.0-current (GENERIC.MP) #166: Wed Feb  8 19:15:03 MST 2017
    deraadt@amd64.openbsd.org:/usr/src/sys/arch/amd64/compile/GENERIC.MP

$ stack --version
Version 1.3.2 x86_64 hpack-0.15.0

$ stack new test82
Downloading template "new-template" to create project "test82" in test82/ ...

The following parameters were needed by the template but not provided: category
You can provide them in /home/user/.stack/config.yaml, like this:
templates:
  params:
    category: value
Or you can pass each one as parameters like this:
stack new test82 new-template -p "category:value"

Looking for .cabal or package.yaml files to use to init the project.
Using cabal packages:
- test82/test82.cabal

Selecting the best among 10 snapshots...

Downloaded lts-8.0 build plan.    
Updating package index Hackage (mirrored at https://github.com/commercialhaskell                                                                                Fetching package index ...remote: Counting objects: 3689, done.
remote: Compressing objects: 100% (246/246), done.
remote: Total 3689 (delta 1313), reused 1262 (delta 1262), pack-reused 2175
Receiving objects: 100% (3689/3689), 2.83 MiB | 725.00 KiB/s, done.
Resolving deltas: 100% (1490/1490), completed with 556 local objects.
From https://github.com/commercialhaskell/all-cabal-hashes
   79f894d3f9..53fcf98366  hackage         -> origin/hackage
 t [tag update]            current-hackage -> current-hackage
Fetched package index.    
Populated index cache.    
* Matches lts-8.0

Selected resolver: lts-8.0
Initialising configuration using resolver: lts-8.0
Total number of user packages considered: 1
Writing configuration to file: test82/stack.yaml
All done.

$ cd test82/

$ stack setup
Preparing to install GHC to an isolated location.
This will not interfere with any system-level installation.
Downloaded ghc-8.0.2.                                     
Unpacking GHC into /home/user/.stack/programs/x86_64-openbsd/ghc-8.0.2.temp/                                                                                 Installed GHC.     
stack will use a sandboxed GHC it installed
For more information on paths, see 'stack path' and 'stack exec env'
To use this GHC and packages outside of a project, consider using:
stack ghc, stack ghci, stack runghc, or stack exec

$ stack build
Warning: File listed in test82.cabal file does not exist: README.md
[1 of 2] Compiling Main             ( /home/user/.stack/setup-exe-src/setup-mPHDZzAJ.hs, /home/user/.stack/setup-exe-src/setup-mPHDZzAJ.o )
[2 of 2] Compiling StackSetupShim   ( /home/user/.stack/setup-exe-src/setup-shim-mPHDZzAJ.hs, /home/user/.stack/setup-exe-src/setup-shim-mPHDZzAJ.o )
Linking /home/user/.stack/setup-exe-cache/x86_64-openbsd/tmp-Cabal-simple_mPHDZzAJ_1.24.2.0_ghc-8.0.2 ...
/home/user/.stack/programs/x86_64-openbsd/ghc-8.0.2/lib/ghc-8.0.2/base-4.9.1.0/libHSbase-4.9.1.0.a(IO__1.o): In function `ghczuwrapperZC0ZCbaseZCSystemziIOZCrand':
(.text+0x1): warning: warning: rand() may return deterministic values, is that what you want?
/home/user/.stack/programs/x86_64-openbsd/ghc-8.0.2/lib/ghc-8.0.2/rts/libHSrts_thr.a(RtsFlags.thr_o): In function `copyArg':

rts/RtsFlags.c:1805:0: error:
     warning: warning: strcpy() is almost always misused, please use strlcpy()
/usr/local/lib/libgmp.so.10.0: warning: warning: vsprintf() is often misused, please use vsnprintf()
/home/user/.stack/programs/x86_64-openbsd/ghc-8.0.2/lib/ghc-8.0.2/rts/libHSrts_thr.a(RtsUtils.thr_o): In function `showStgWord64':

rts/RtsUtils.c:230:0: error:
     warning: warning: sprintf() is often misused, please use snprintf()
test82-0.1.0.0: configure (lib + exe)
Configuring test82-0.1.0.0...
test82-0.1.0.0: build (lib + exe)
Preprocessing library test82-0.1.0.0...
[1 of 1] Compiling Lib              ( src/Lib.hs, .stack-work/dist/x86_64-openbsd/Cabal-1.24.2.0/build/Lib.o )
Preprocessing executable 'test82-exe' for test82-0.1.0.0...
[1 of 1] Compiling Main             ( app/Main.hs, .stack-work/dist/x86_64-openbsd/Cabal-1.24.2.0/build/test82-exe/test82-exe-tmp/Main.o )
Linking .stack-work/dist/x86_64-openbsd/Cabal-1.24.2.0/build/test82-exe/test82-exe ...
/home/user/.stack/programs/x86_64-openbsd/ghc-8.0.2/lib/ghc-8.0.2/rts/libHSrts_thr.a(RtsFlags.thr_o): In function `copyArg':

rts/RtsFlags.c:1805:0: error:
     warning: warning: strcpy() is almost always misused, please use strlcpy()
/usr/local/lib/libgmp.so.10.0: warning: warning: vsprintf() is often misused, please use vsnprintf()
/home/user/.stack/programs/x86_64-openbsd/ghc-8.0.2/lib/ghc-8.0.2/rts/libHSrts_thr.a(RtsUtils.thr_o): In function `showStgWord64':

rts/RtsUtils.c:230:0: error:
     warning: warning: sprintf() is often misused, please use snprintf()
Warning: File listed in test82.cabal file does not exist: README.md
test82-0.1.0.0: copy/register
Installing library in
/home/user/Scratch/test82/.stack-work/install/x86_64-openbsd/lts-8.0/8.0.2/lib/x86_64-openbsd-ghc-8.0.2/test82-0.1.0.0-EVLQkMVlEaDLN2y26yqY67
Installing executable(s) in
/home/user/Scratch/test82/.stack-work/install/x86_64-openbsd/lts-8.0/8.0.2/bin
Registering test82-0.1.0.0...

$ stack exec test82-exe
someFunc
```

Thanks @adamsteen,
A little more on [OpenBSD's new wxallowed attribute](https://www.openbsd.org/faq/upgrade60.html). You can either set it by editing the `/etc/fstab` like so:
`01020304050607.h /usr/local ffs rw,nodev,wxallowed 1 2` and rebooting, or you can temporary set a mount as wxallowed by running this command: `mount -uo wxallowed /usr/local.` (where /usr/local is the partition you'd like to affected, run `mount` to see them all)

In either case, I'm still having trouble getting ghc-8.0 to compile on OpenBSD 6.0. Here's the config and error I keep receiving after trying `gmake`:

```
Configure completed successfully.

   Building GHC version  : 8.0.2.20170314
          Git commit id  : d2d13a4f6750e30389552974bd7465712d522735

   Build platform        : x86_64-unknown-openbsd
   Host platform         : x86_64-unknown-openbsd
   Target platform       : x86_64-unknown-openbsd

   Bootstrapping using   : /usr/local/bin/ghc
      which is version   : 7.10.3

   Using gcc                 : /usr/bin/gcc
      which is version       : 4.2.1
   Building a cross compiler : NO
   hs-cpp       : /usr/bin/gcc
   hs-cpp-flags : -E -undef -traditional
   ld           : /usr/bin/ld
   nm           : /usr/bin/nm
   objdump      : /usr/bin/objdump
   Happy        : /home/goldfish/.cabal/bin/happy (1.19.5)
   Alex         : /home/goldfish/.cabal/bin/alex (3.2.1)
   Perl         : /usr/bin/perl
   sphinx-build :
   xelatex      :

   Using LLVM tools
      llc   :
      opt   :

   HsColour was not found; documentation will not contain source links

   Tools to build Sphinx HTML documentation available: NO
   Tools to build Sphinx PDF documentation available: NO
----------------------------------------------------------------------

For a standard build of GHC (fully optimised with profiling), type (g)make.

To make changes to the default build configuration, copy the file
mk/build.mk.sample to mk/build.mk, and edit the settings in there.

For more information on how to configure your GHC build, see
   http://ghc.haskell.org/trac/ghc/wiki/Building
```
```
$ gmake
===--- building phase 0
gmake --no-print-directory -f ghc.mk phase=0 phase_0_builds
gmake[1]: Nothing to be done for 'phase_0_builds'.
===--- building phase 1
gmake --no-print-directory -f ghc.mk phase=1 phase_1_builds
gmake[1]: Nothing to be done for 'phase_1_builds'.
===--- building final phase
gmake --no-print-directory -f ghc.mk phase=final all
Panic! libraries/Cabal/Cabal/dist-install/build/Distribution/Simple/Setup.hi exists, but libraries/Cabal/Cabal/dist-install/build/Distribution/Simple/Setup.dyn_o does not.
gmake[1]: *** [libraries/Cabal/Cabal/ghc.mk:4: libraries/Cabal/Cabal/dist-install/build/Distribution/Simple/Setup.dyn_o] Error 1
gmake: *** [Makefile:134: all] Error 2
```

also, I've been trying to build the latest master, but each time I get some different error, but maybe I should expect that from the latest. Here's the config and error message I receive when trying to build the latest master as of April 9, 2017:

```
Configure completed successfully.

   Building GHC version  : 8.3.20170409
          Git commit id  : 87377f74eec1567af741737b4b9034d06e3f0698

   Build platform        : x86_64-unknown-openbsd
   Host platform         : x86_64-unknown-openbsd
   Target platform       : x86_64-unknown-openbsd

   Bootstrapping using   : /usr/local/bin/ghc
      which is version   : 7.10.3

   Using (for bootstrapping) : /usr/bin/gcc
   Using gcc                 : gcc
      which is version       : 4.2.1
   Building a cross compiler : NO
   Unregisterised            : NO
   hs-cpp       : gcc
   hs-cpp-flags : -E -undef -traditional
   ld           : /usr/bin/ld
   nm           : /usr/bin/nm
   objdump      : /usr/bin/objdump
   Happy        : /home/goldfish/.cabal/bin/happy (1.19.5)
   Alex         : /home/goldfish/.cabal/bin/alex (3.2.1)
   Perl         : /usr/bin/perl
   sphinx-build :
   xelatex      :

   Using LLVM tools
      llc   :
      opt   :

   HsColour was not found; documentation will not contain source links

   Tools to build Sphinx HTML documentation available: NO
   Tools to build Sphinx PDF documentation available: NO
----------------------------------------------------------------------

For a standard build of GHC (fully optimised with profiling), type (g)make.

To make changes to the default build configuration, copy the file
mk/build.mk.sample to mk/build.mk, and edit the settings in there.

For more information on how to configure your GHC build, see
   http://ghc.haskell.org/trac/ghc/wiki/Building
```
```
$ gmake
===--- building phase 0
gmake --no-print-directory -f ghc.mk phase=0 phase_0_builds
gmake[1]: Nothing to be done for 'phase_0_builds'.
===--- building phase 1
gmake --no-print-directory -f ghc.mk phase=1 phase_1_builds
gcc -E -undef -traditional -P -DTOP='"/home/goldfish/yesod_test/ghc_compile_test/ghc"' -DFFI_INCLUDE_DIR=/usr/local/include -DFFI_LIB_DIR=/usr/local/lib '-DFFI_LIB=' -x c -Iincludes -Iincludes/dist -Iincludes/dist-derivedconstants/header -Iincludes/dist-ghcconstants/header rts/package.conf.in -o rts/dist/package.conf.inplace.raw
grep -v '^#pragma GCC' rts/dist/package.conf.inplace.raw | sed -e 's/""//g' -e 's/:[    ]*,/: /g' > rts/dist/package.conf.inplace
"inplace/bin/ghc-pkg" update --force rts/dist/package.conf.inplace
ghc-pkg: Couldn't open database /home/goldfish/yesod_test/ghc_compile_test/ghc/inplace/lib/package.conf.d for modification: /home/goldfish/yesod_test/ghc_compile_test/ghc/inplace/lib/package.conf.d/package.cache: GHC.PackageDb.readPackageDb: inappropriate type (Not a valid Unicode code point!)
gmake[1]: *** [rts/ghc.mk:540: rts/dist/package.conf.inplace] Error 1
gmake[1]: *** Deleting file 'rts/dist/package.conf.inplace'
gmake: *** [Makefile:125: all] Error 2
$
```

If anyone ha seen these errors before and can point me in the right direction that would be very much appreciated.

First, building GHC can be hard in general, also on Linux, even if you know
the Makefile language, and even if you read all docs on the GHC wiki. I
don't want to discourage you, but please read those docs. There are enough
bugs to workaround.

This sort of error

Panic! libraries/Cabal/Cabal/dist-install/build/Distribution/Simple/Setup.hi
exists, but libraries/Cabal/Cabal/dist-install/build/Distribution/Simple/Setup.dyn_o
does not.


is often produced by pressing Ctrl-C after starting a build, so that only
some of the output from one input are saved (here, the .hi, but not
.dyn_o). This is a GHC bug and also happens during builds with stack or
cabal. It happens to be often enough.
As a workaround, remove the output which is there.

If you get that on a clean build (seems unlikely here): maybe the dyn_o
file is just not being created. That's output for dynamic linking, but
maybe it doesn't work on OpenBSD. Maybe the build.mk is not set correctly.

The second error is weirder. If you get it on a clean build, check
permissions of the file where the build is failing.
-- 
Paolo Giarrusso
From smartphone, sorry for typos or excessive brevity


Il 10 apr 2017 03:30, "Sean Westfall" <notifications@github.com> ha scritto:

Thanks @adamsteen <https://github.com/adamsteen>,
A little more on OpenBSD's new wxallowed attribute
<https://www.openbsd.org/faq/upgrade60.html>. You can either set it by
editing the /etc/fstab like so:
01020304050607.h /usr/local ffs rw,nodev,wxallowed 1 2 and rebooting, or
you can temporary set a mount as wxallowed by running this command: mount
-uo wxallowed /usr/local. (where /usr/local is the partition you'd like to
affected, run mount to see them all)

In either case, I'm still having trouble getting ghc-8.0 to compile on
OpenBSD 6.0. Here's the config and error I keep receiving after trying gmake
:

Configure completed successfully.

   Building GHC version  : 8.0.2.20170314
          Git commit id  : d2d13a4f6750e30389552974bd7465712d522735

   Build platform        : x86_64-unknown-openbsd
   Host platform         : x86_64-unknown-openbsd
   Target platform       : x86_64-unknown-openbsd

   Bootstrapping using   : /usr/local/bin/ghc
      which is version   : 7.10.3

   Using gcc                 : /usr/bin/gcc
      which is version       : 4.2.1
   Building a cross compiler : NO
   hs-cpp       : /usr/bin/gcc
   hs-cpp-flags : -E -undef -traditional
   ld           : /usr/bin/ld
   nm           : /usr/bin/nm
   objdump      : /usr/bin/objdump
   Happy        : /home/goldfish/.cabal/bin/happy (1.19.5)
   Alex         : /home/goldfish/.cabal/bin/alex (3.2.1)
   Perl         : /usr/bin/perl
   sphinx-build :
   xelatex      :

   Using LLVM tools
      llc   :
      opt   :

   HsColour was not found; documentation will not contain source links

   Tools to build Sphinx HTML documentation available: NO
   Tools to build Sphinx PDF documentation available: NO
----------------------------------------------------------------------

For a standard build of GHC (fully optimised with profiling), type (g)make.

To make changes to the default build configuration, copy the file
mk/build.mk.sample to mk/build.mk, and edit the settings in there.

For more information on how to configure your GHC build, see
   http://ghc.haskell.org/trac/ghc/wiki/Building

$ gmake
===--- building phase 0
gmake --no-print-directory -f ghc.mk phase=0 phase_0_builds
gmake[1]: Nothing to be done for 'phase_0_builds'.
===--- building phase 1
gmake --no-print-directory -f ghc.mk phase=1 phase_1_builds
gmake[1]: Nothing to be done for 'phase_1_builds'.
===--- building final phase
gmake --no-print-directory -f ghc.mk phase=final all
Panic! libraries/Cabal/Cabal/dist-install/build/Distribution/Simple/Setup.hi
exists, but libraries/Cabal/Cabal/dist-install/build/Distribution/Simple/Setup.dyn_o
does not.
gmake[1]: *** [libraries/Cabal/Cabal/ghc.mk:4:
libraries/Cabal/Cabal/dist-install/build/Distribution/Simple/Setup.dyn_o]
Error 1
gmake: *** [Makefile:134: all] Error 2

also, I've been trying to build the latest master, but each time I get some
different error, but maybe I should expect that from the latest. Here's the
config and error message I receive when trying to build the latest master
as of April 9, 2017:

Configure completed successfully.

   Building GHC version  : 8.3.20170409
          Git commit id  : 87377f74eec1567af741737b4b9034d06e3f0698

   Build platform        : x86_64-unknown-openbsd
   Host platform         : x86_64-unknown-openbsd
   Target platform       : x86_64-unknown-openbsd

   Bootstrapping using   : /usr/local/bin/ghc
      which is version   : 7.10.3

   Using (for bootstrapping) : /usr/bin/gcc
   Using gcc                 : gcc
      which is version       : 4.2.1
   Building a cross compiler : NO
   Unregisterised            : NO
   hs-cpp       : gcc
   hs-cpp-flags : -E -undef -traditional
   ld           : /usr/bin/ld
   nm           : /usr/bin/nm
   objdump      : /usr/bin/objdump
   Happy        : /home/goldfish/.cabal/bin/happy (1.19.5)
   Alex         : /home/goldfish/.cabal/bin/alex (3.2.1)
   Perl         : /usr/bin/perl
   sphinx-build :
   xelatex      :

   Using LLVM tools
      llc   :
      opt   :

   HsColour was not found; documentation will not contain source links

   Tools to build Sphinx HTML documentation available: NO
   Tools to build Sphinx PDF documentation available: NO
----------------------------------------------------------------------

For a standard build of GHC (fully optimised with profiling), type (g)make.

To make changes to the default build configuration, copy the file
mk/build.mk.sample to mk/build.mk, and edit the settings in there.

For more information on how to configure your GHC build, see
   http://ghc.haskell.org/trac/ghc/wiki/Building

$ gmake
===--- building phase 0
gmake --no-print-directory -f ghc.mk phase=0 phase_0_builds
gmake[1]: Nothing to be done for 'phase_0_builds'.
===--- building phase 1
gmake --no-print-directory -f ghc.mk phase=1 phase_1_builds
gcc -E -undef -traditional -P
-DTOP='"/home/goldfish/yesod_test/ghc_compile_test/ghc"'
-DFFI_INCLUDE_DIR=/usr/local/include -DFFI_LIB_DIR=/usr/local/lib
'-DFFI_LIB=' -x c -Iincludes -Iincludes/dist
-Iincludes/dist-derivedconstants/header
-Iincludes/dist-ghcconstants/header rts/package.conf.in -o
rts/dist/package.conf.inplace.raw
grep -v '^#pragma GCC' rts/dist/package.conf.inplace.raw | sed -e
's/""//g' -e 's/:[    ]*,/: /g' > rts/dist/package.conf.inplace
"inplace/bin/ghc-pkg" update --force rts/dist/package.conf.inplace
ghc-pkg: Couldn't open database
/home/goldfish/yesod_test/ghc_compile_test/ghc/inplace/lib/package.conf.d
for modification:
/home/goldfish/yesod_test/ghc_compile_test/ghc/inplace/lib/package.conf.d/package.cache:
GHC.PackageDb.readPackageDb: inappropriate type (Not a valid Unicode
code point!)
gmake[1]: *** [rts/ghc.mk:540: rts/dist/package.conf.inplace] Error 1
gmake[1]: *** Deleting file 'rts/dist/package.conf.inplace'
gmake: *** [Makefile:125: all] Error 2
$

If anyone ha seen these errors before and can point me in the right
direction that would be very much appreciated.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/commercialhaskell/stack/issues/416#issuecomment-292829292>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AARsqHFqHruvGho3kc8xvCRjI4ovgqWVks5ruYZSgaJpZM4FL9qe>
.

--- a/src/Stack/Setup.hs
+++ b/src/Stack/Setup.hs
@@ -33,6 +33,7 @@ module Stack.Setup
   , downloadStackExe
   ) where
 
+import              Bindings.Uname (uname, release)
 import qualified    Codec.Archive.Tar as Tar
 import              Control.Applicative (empty)
 import              Control.Monad.Logger
@@ -52,6 +53,7 @@ import              Data.Foldable (maximumBy)
 import qualified    Data.HashMap.Strict as HashMap
 import              Data.IORef.RunOnce (runOnce)
 import              Data.List hiding (concat, elem, maximumBy, any)
+import              Data.List.Split (splitOn)
 import qualified    Data.Map as Map
 import qualified    Data.Set as Set
 import qualified    Data.Text as T
@@ -59,9 +61,11 @@ import qualified    Data.Text.Encoding as T
 import qualified    Data.Text.Encoding.Error as T
 import              Data.Time.Clock (NominalDiffTime, diffUTCTime, getCurrentTime)
 import qualified    Data.Yaml as Yaml
-import              Distribution.System (OS (Linux), Arch (..), Platform (..))
+import              Distribution.System (OS (Linux, OpenBSD), Arch (..), Platform (..))
 import qualified    Distribution.System as Cabal
 import              Distribution.Text (simpleParse)
+import              Foreign.C (throwErrnoIfMinus1_, peekCString)
+import              Foreign.Marshal (alloca)
 import              Lens.Micro (set)
 import              Network.HTTP.Simple (getResponseBody, httpLBS, withResponse, getResponseStatusCode)
 import              Network.HTTP.Download
@@ -601,6 +605,9 @@ getGhcBuild menv = do
                 case libComponents ++ pieComponents of
                     [] -> useBuild CompilerBuildStandard
                     components -> useBuild (CompilerBuildSpecialized (intercalate "-" components))
+            Platform _ Cabal.OpenBSD -> do
+                releaseStr <- mungeRelease <$> sysRelease
+                useBuild (CompilerBuildSpecialized releaseStr)
             _ -> useBuild CompilerBuildStandard
     useBuild CompilerBuildStandard = do
         $logDebug "Using standard GHC build"
@@ -608,6 +615,24 @@ getGhcBuild menv = do
     useBuild (CompilerBuildSpecialized s) = do
         $logDebug ("Using " <> T.pack s <> " GHC build")
         return (CompilerBuildSpecialized s)
+
+-- | Encode an OpenBSD version (like "6.1") into a valid argument for
+-- CompilerBuildSpecialized, so "maj6-min1". Later version numbers are prefixed
+-- with "r".
+-- The result r must be such that "ghc-" ++ r is a valid package name,
+-- as recognized by parsePackageNameFromString.
+mungeRelease :: String -> String
+mungeRelease = intercalate "-" . prefixMaj . splitOn "."
+  where
+    prefixFst pfx k (rev : revs) = (pfx ++ rev) : k revs
+    prefixFst _ _ [] = []
+    prefixMaj = prefixFst "maj" prefixMin
+    prefixMin = prefixFst "min" (map ('r':))
+
+sysRelease :: MonadIO m => m String
+sysRelease = liftIO $ alloca $ \ ptr ->
+           do throwErrnoIfMinus1_ "uname" $ uname ptr
+              peekCString $ release ptr
 
 -- | Ensure Docker container-compatible 'stack' executable is downloaded
 ensureDockerStackExe :: HasConfig env => Platform -> RIO env (Path Abs File)

--- a/src/Stack/Setup.hs
+++ b/src/Stack/Setup.hs
@@ -33,7 +33,6 @@ module Stack.Setup
   , downloadStackExe
   ) where
 
-import              Bindings.Uname (uname, release)
 import qualified    Codec.Archive.Tar as Tar
 import              Control.Applicative (empty)
 import              Control.Monad.Logger
@@ -64,8 +63,6 @@ import qualified    Data.Yaml as Yaml
 import              Distribution.System (OS, Arch (..), Platform (..))
 import qualified    Distribution.System as Cabal
 import              Distribution.Text (simpleParse)
-import              Foreign.C (throwErrnoIfMinus1_, peekCString)
-import              Foreign.Marshal (alloca)
 import              Lens.Micro (set)
 import              Network.HTTP.Simple (getResponseBody, httpLBS, withResponse, getResponseStatusCode)
 import              Network.HTTP.Download
@@ -109,7 +106,10 @@ import              System.Process.Run (runCmd, Cmd(..))
 import              Text.Printf (printf)
 
 #if !WINDOWS
-import           System.Posix.Files (setFileMode)
+import              Bindings.Uname (uname, release)
+import              Foreign.C (throwErrnoIfMinus1_, peekCString)
+import              Foreign.Marshal (alloca)
+import              System.Posix.Files (setFileMode)
 #endif
 
 -- | Default location of the stack-setup.yaml file
@@ -605,9 +605,11 @@ getGhcBuild menv = do
                 case libComponents ++ pieComponents of
                     [] -> useBuild CompilerBuildStandard
                     components -> useBuild (CompilerBuildSpecialized (intercalate "-" components))
+#if !WINDOWS
             Platform _ Cabal.OpenBSD -> do
                 releaseStr <- mungeRelease <$> sysRelease
                 useBuild (CompilerBuildSpecialized releaseStr)
+#endif
             _ -> useBuild CompilerBuildStandard
     useBuild CompilerBuildStandard = do
         $logDebug "Using standard GHC build"
@@ -629,6 +631,7 @@ mungeRelease = intercalate "-" . prefixMaj . splitOn "."
     prefixMaj = prefixFst "maj" prefixMin
     prefixMin = prefixFst "min" (map ('r':))
 
+#if !WINDOWS
 sysRelease :: (MonadUnliftIO m, MonadLogger m) => m String
 sysRelease =
   handleIO (\e -> do
@@ -640,6 +643,7 @@ sysRelease =
   alloca $ \ ptr ->
              do throwErrnoIfMinus1_ "uname" $ uname ptr
                 peekCString $ release ptr
+#endif
 
 -- | Ensure Docker container-compatible 'stack' executable is downloaded
 ensureDockerStackExe :: HasConfig env => Platform -> RIO env (Path Abs File)

--- a/src/Stack/Setup.hs
+++ b/src/Stack/Setup.hs
@@ -61,7 +61,7 @@ import qualified    Data.Text.Encoding as T
 import qualified    Data.Text.Encoding.Error as T
 import              Data.Time.Clock (NominalDiffTime, diffUTCTime, getCurrentTime)
 import qualified    Data.Yaml as Yaml
-import              Distribution.System (OS (Linux, OpenBSD), Arch (..), Platform (..))
+import              Distribution.System (OS, Arch (..), Platform (..))
 import qualified    Distribution.System as Cabal
 import              Distribution.Text (simpleParse)
 import              Foreign.C (throwErrnoIfMinus1_, peekCString)
@@ -546,7 +546,7 @@ getGhcBuild menv = do
 
         platform <- view platformL
         case platform of
-            Platform _ Linux -> do
+            Platform _ Cabal.Linux -> do
                 -- Some systems don't have ldconfig in the PATH, so make sure to look in /sbin and /usr/sbin as well
                 sbinEnv <- modifyEnvOverride menv $
                     Map.insert "PATH" $

--- a/src/Stack/Setup.hs
+++ b/src/Stack/Setup.hs
@@ -629,10 +629,17 @@ mungeRelease = intercalate "-" . prefixMaj . splitOn "."
     prefixMaj = prefixFst "maj" prefixMin
     prefixMin = prefixFst "min" (map ('r':))
 
-sysRelease :: MonadIO m => m String
-sysRelease = liftIO $ alloca $ \ ptr ->
-           do throwErrnoIfMinus1_ "uname" $ uname ptr
-              peekCString $ release ptr
+sysRelease :: (MonadUnliftIO m, MonadLogger m) => m String
+sysRelease =
+  handleIO (\e -> do
+               $logWarn $ T.concat [ T.pack "Could not query OS version"
+                                   , T.pack $ show e
+                                   ]
+               return "") .
+  liftIO .
+  alloca $ \ ptr ->
+             do throwErrnoIfMinus1_ "uname" $ uname ptr
+                peekCString $ release ptr
 
 -- | Ensure Docker container-compatible 'stack' executable is downloaded
 ensureDockerStackExe :: HasConfig env => Platform -> RIO env (Path Abs File)

--- a/stack-7.10.yaml
+++ b/stack-7.10.yaml
@@ -39,3 +39,4 @@ extra-deps:
 - unliftio-core-0.1.0.0
 - unliftio-0.1.0.0
 - async-2.1.1.1
+- bindings-uname-0.1

--- a/stack-nightly.yaml
+++ b/stack-nightly.yaml
@@ -15,3 +15,4 @@ flags:
 extra-deps:
 - mintty-0.1.1
 - text-metrics-0.3.0
+- bindings-uname-0.1

--- a/stack.cabal
+++ b/stack.cabal
@@ -195,6 +195,7 @@ library
                    , base >= 4.8 && <5
                    , base64-bytestring
                    , binary >= 0.7 && < 0.9
+                   , bindings-uname >= 0.1
                    , binary-tagged >= 0.1.1
                    , blaze-builder
                    , bytestring >= 0.10.4.0

--- a/stack.cabal
+++ b/stack.cabal
@@ -195,7 +195,6 @@ library
                    , base >= 4.8 && <5
                    , base64-bytestring
                    , binary >= 0.7 && < 0.9
-                   , bindings-uname >= 0.1
                    , binary-tagged >= 0.1.1
                    , blaze-builder
                    , bytestring >= 0.10.4.0
@@ -277,6 +276,7 @@ library
   else
     build-depends:   unix >= 2.7.0.1
                    , pid1 >= 0.1 && < 0.2
+                   , bindings-uname >= 0.1
   default-language:  Haskell2010
 
 executable stack

--- a/stack.yaml
+++ b/stack.yaml
@@ -22,3 +22,4 @@ extra-deps:
 - unicode-transforms-0.3.2
 - unliftio-core-0.1.0.0
 - unliftio-0.1.0.0
+- bindings-uname-0.1

